Support import from source (#3856)
mlpack/mlpack: CMakeLists.txt

cmake_minimum_required(VERSION 3.6)
project(mlpack C CXX)

> #Replace all CMAKE_SOURCE_DIR and CMAKE_BINARY_DIR with MLPACK_ prefix.

Reviewer: ```suggestion
# We use `MLPACK_SOURCE_DIR` and `MLPACK_BINARY_DIR` throughout for the
# situation where mlpack's CMake configuration is being included by another
# project.
```
Just a simple clarification; let me know what you think.

set(MLPACK_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} CACHE STRING "Source path of mlpack")

> set(MLPACK_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR} CACHE STRING "Binary path of mlpack")

Reviewer: Does it need to be a cache variable?  I had thought that cache variables were really meant to be user-configurable, but if the user specifies something random for these, I don't think it will work very well---`MLPACK_SOURCE_DIR` and `MLPACK_BINARY_DIR` need to be the root of the mlpack CMake configuration.

  set(MLPACK_VERSION_STRING "@MLPACK_VERSION_STRING@")
  configure_file(

>     ${CMAKE_CURRENT_SOURCE_DIR}/CMake/mlpack.pc.in

Reviewer: ```suggestion
    ${MLPACK_SOURCE_DIR}/CMake/mlpack.pc.in
```
(small change, might as well change this too for consistency)

    ${MLPACK_BINARY_DIR}/CMake/mlpack.pc.in.partial @ONLY)

  add_custom_target(pkgconfig ALL
      ${CMAKE_COMMAND}
          -D MLPACK_SOURCE_DIR="${MLPACK_SOURCE_DIR}"
          -D MLPACK_BINARY_DIR="${MLPACK_BINARY_DIR}"

>           -P "${CMAKE_CURRENT_SOURCE_DIR}/CMake/GeneratePkgConfig.cmake"

Reviewer: ```suggestion
          -P "${MLPACK_SOURCE_DIR}/CMake/GeneratePkgConfig.cmake"
```
(another small change that I saw)
